src/Message/ReversalResponse.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Omnipay\MOLPay\Message;
+
+use Omnipay\Common\Message\AbstractResponse;
+use Omnipay\Common\Message\RequestInterface;
+
+/**
+ * Class ReversalResponse
+ * @package Omnipay\MOLPay\Message
+ *
+ * MOLPay Reversal Response
+ *
+ * * TranID     - Unique transaction ID for tracking purpose
+ * * Domain     - Merchant ID in MOLPay system
+ * * VrfKey     - This is the data integrity protection hash string
+ * * StatCode   - 00 = Success
+ *                11 = Failure
+ *                12 = Invalid or unmatched security hash string
+ *                13 = Not a refundable transaction
+ *                14 = Transaction date more than 45 days
+ *                15 = Requested day is on settlement day
+ *                16 = Forbidden transaction
+ *                17 = Transaction not found
+ *
+ * * StatDate   - Response date & time
+ *
+ */
+class ReversalResponse extends AbstractResponse
+{
+    /**
+     * Reversal Request Response statCodes and messages
+     */
+    protected $statCodeMessages = array(
+        '00' => 'Success',
+        '11' => 'Failure',
+        '12' => 'Invalid or unmatched security hash string',
+        '13' => 'Not a refundable transaction',
+        '14' => 'Transaction date more than 45 days',
+        '15' => 'Requested day is on settlement day',
+        '16' => 'Forbidden transaction',
+        '17' => 'Transaction not found'
+    );
+
+
+    /**
+     * ReversalResponse constructor.
+     * @param RequestInterface $request
+     * @param mixed $data
+     */
+    public function __construct(RequestInterface $request, $data)
+    {
+        $this->request = $request;
+
+        $search = array("\r\n", "\n", "\r");
+        $data = str_replace($search, '&', $data);
```

Comment:
Maybe we can use this? https://github.com/thephpleague/omnipay-common/blob/master/src/Common/Http/ResponseParser.php

Reply:
This is because MOLPay returns Content-Type: text/plain; charset=utf-8 by default in reversal response, these lines are parsing the text file and populate into a key-value pair array.
See page 73 Reversal Request parameter 'type', the default is plain text result. 

Besides, you may compare the unit test mock(which is the edited live response from MOLPay endpoint ) ReversalSuccess.txt(plain text result) and PartialRefundPending.txt(json response).